### ISSUE TITLE
nn.ConvTranspose3d with output_padding

### ISSUE THREAD

**Issue by azad96** [NVIDIA/TensorRT#2168]
Hi everyone,

I have an input of size [1, 64, 6, 15, 20], and when I apply torch's ConvTranspose3d on it with the following arguments

ConvTranspose3d(64, 32, kernel_size=(3, 3, 3), stride=(2, 2, 2), padding=(1, 1, 1), output_padding=(0, 0, 0), bias=False)

I get an output of size [1, 32, 11, 29, 39]

I managed to get the same output in TensorRT with IDeconvolutionLayer as follows:

auto deconv1 = network->addDeconvolutionNd(*input, 32, Dims{3, {3, 3, 3}}, deconvWeights, emptyWeights);
assert(deconv1);
deconv1->setStrideNd(nvinfer1::Dims{3, {2, 2, 2}});
deconv1->setPaddingNd(nvinfer1::Dims{3, {1, 1, 1}});

However, I have a ConvTranspose3d layer with output_padding=(1, 1, 1), which results in an output of size [1, 32, 12, 30, 40]. IDeconvolutionLayer has some methods namely setPrePadding, setPostPadding, setPaddingMode. I tried using these to get the same output shape but they did not make any difference.

My question is whether output_padding is supported in TensorRT and if so how to use it?

**Comment by zerollzeng** [comment]
https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_convolution_layer.html

We don't support padding output for an IConvolutionLayer or IDeconvolution layer. prePadding and postPadding is used for asymmetric padding value. e.g. In a 2d conv, your want to pad the input with (1,1) on top and left. and (2,2) on the right and bottom. @ttyio correct me if I'm wrong

But I think you can add an IPadding layer(https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_padding_layer.html) to your output.

**Comment by azad96** [comment]
@zerollzeng I will check the IPadding layer and try to see if the outputs match. thank you for the idea

**Comment by azad96** [comment]
@zerollzeng I tried adding a IPadding layer but only for 2d padding is supported, so I can only add padding to the last two dimensions.

**Comment by zerollzeng** [comment]
Sorry, I miss the limitation, it should be feasible to do it with ISliceLayer(https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_slice_layer.html).

Try adding a slice layer with a negative size.

**Comment by azad96** [comment]
thanks again, I will give it a go and inform you as well

**Comment by azad96** [comment]
@zerollzeng , could you elaborate more on how to use a slice layer with a negative size. I tried a couple of things similar to the below code

~~~
// input shape: [32, 11, 29, 39]
auto slice1 = network->addSlice(*input, 
                                nvinfer1::Dims{4, {0, 0, 0, 0}},
                                nvinfer1::Dims{4, {32, 12, 30, 40}},
                                nvinfer1::Dims{4, {1, -1, -1, -1}});
assert(slice1);
// slice1->setMode(nvinfer1::SliceMode::kFILL);
~~~

I am not sure how to configure the stride and I guess I should also set a slice mode. Currently, I am getting the following error

~~~
ERROR: 4: (Unnamed Layer* 22) [Slice]: out of bounds slice, input dimensions = [32,11,29,39], start = [0,0,0,0], size = [32,12,30,40], stride = [1,-1,-1,-1].
ERROR: 4: [network.cpp::validate::2563] Error Code 4: Internal Error (Layer (Unnamed Layer* 22) [Slice] failed validation)
ERROR: 2: [builder.cpp::buildSerializedNetwork::399] Error Code 2: Internal Error (Assertion enginePtr != nullptr failed.)
cost_reg_net: /home/azad/Desktop/TensorRT-API-New/costRegNet/src/cost_reg_net.cpp:145: virtual void CostRegNet::SerializeEngine(char*, std::map<std::__cxx11::basic_string<char>, nvinfer1::Weights>&, nvinfer1::IBuilder*&, nvinfer1::IBuilderConfig*&, nvinfer1::IHostMemory*&, unsigned int, nvinfer1::DataType): Assertion `modelStream != nullptr' failed.
~~~

**Comment by zerollzeng** [comment]
I am not an expert on this, perhaps you can refer to https://github.com/onnx/onnx-tensorrt/blob/c3cfcbc8248c6bd007e6630af2085df5e4834b42/builtin_op_importers.cpp#L2972. that's how we parse an onnx Pad node to TensorRT. @pranavm-nvidia @kevinch-nv might know more about this.

**Comment by pranavm-nvidia** [comment]
@azad96 Here's a short Python-based example I had written a while ago. The same idea should apply with the C++ API:
```py
#!/usr/bin/env python3
# Generation Command: polygraphy template trt-network -o slice_padding.py
from polygraphy import func
from polygraphy.backend.trt import CreateNetwork
import tensorrt as trt

"""
Run with:
polygraphy run slice_padding.py --trt

To see output tensors, use:
polygraphy run slice_padding.py --trt -vv --val-range [1,1]
"""


@func.extend(CreateNetwork())
def load_network(builder, network):
    inp = network.add_input("input", shape=(1, 1, 2, 2, 2), dtype=trt.float32)

    # We can use a slice layer to apply padding by slicing out-of-bounds.
    # Set the slice mode to either trt.SliceMode.FILL or trt.SliceMode.REFLECT to implement
    # constant padding or reflect padding respectively.

    # Apply a pre-padding of 2 and post-padding of 1 - output shape is (1, 1, 5, 5, 5).
    # To understand this better, consider an example where we look at a 1D tensor of shape (2, ).
    # The contents of this tensor are:
    #       [a, b]
    # The padding amounts are determined by the 'start' and 'shape' arguments like so:
    #       .      .      [a,  b]    .
    #       ^      ^      ^    ^     ^
    #    start=-2  -1     0    1     2
    #      |-------------------------| <- shape=5

    padding = network.add_slice(inp, start=(0, 0, -2, -2, -2), shape=(1, 1, 5, 5, 5), stride=(1, 1, 1, 1, 1))

    padding.mode = trt.SliceMode.FILL
    # Note: The fill value (default = 0) can be controlled by setting the 5th input:
    # padding.set_input(4, <fill_value_tensor>)

    network.mark_output(padding.get_output(0))
```

**Comment by azad96** [comment]
Firstly, thank you for the useful link. Actually, by using kWRAP I can get an output with the correct shape as follows

~~~
auto slice1 = network->addSlice(*deconv1->getOutput(0), 
                                nvinfer1::Dims{4, {0, 0, 0, 0}},
                                nvinfer1::Dims{4, {dims[0], dims[1]+1, dims[2]+1, dims[3]+1}},
                                nvinfer1::Dims{4, {1, 1, 1, 1}});

slice1->setMode(nvinfer1::SliceMode::kWRAP);
~~~

However, I could not use kFILL. I tried the following  but I guess I am creating the constantLayer wrong

~~~
const float fill_value[1] = {0};
auto fill_value_layer = network->addConstant(nvinfer1::Dims{0, {0}}, nvinfer1::Weights{nvinfer1::DataType::kFLOAT, fill_value, 1});
// auto fill_value_layer = network->addConstant(nvinfer1::Dims{1, {1}}, nvinfer1::Weights{nvinfer1::DataType::kFLOAT, fill_value, 1});
auto slice1 = network->addSlice(*deconv1->getOutput(0), 
                                nvinfer1::Dims{4, {0, 0, 0, 0}},
                                nvinfer1::Dims{4, {dims[0], dims[1]+1, dims[2]+1, dims[3]+1}},
                                nvinfer1::Dims{4, {1, 1, 1, 1}});

slice1->setMode(nvinfer1::SliceMode::kFILL);
slice1->setInput(4, *fill_value_layer->getOutput(0));
~~~

I get the error below
![image](https://user-images.githubusercontent.com/39907647/181357548-27dafa4e-9a88-49e0-9c48-343a515c6d5f.png)

**Comment by pranavm-nvidia** [comment]
Which version of TensorRT is this? Is it possible you have a mismatch between the headers and libraries? The first error in your screenshot suggests that TRT doesn't recognize the slice mode, but since your code is compiling, it must be in the header at least.

**Comment by azad96** [comment]
@pranavm-nvidia it's TensorRT 8.0. there might be a mismatch between the headers and libraries. I will look into it

**Comment by ttyio** [comment]
closing since no activity for more than 14 days, please reopen if you still have question, thanks!
